tests/csapi/sync_test.go

```diff
@@ -374,6 +377,188 @@ func TestSync(t *testing.T) {
 	})
 }
 
+// This is a regression test for
+// https://github.com/matrix-org/synapse/issues/16463
+//
+// We test this by having a local user (alice) and remote user (charlie) in a
+// room. Charlie sends 50+ messages into the room without sending to Alice's
+// server. Charlie then sends one more which get sent to Alice.
+//
+// Alice should observer that she receives some (though not all) of charlie's
+// events, with the `limited` flag set.
+func TestSyncTimelineGap(t *testing.T) {
+	runtime.SkipIf(t, runtime.Dendrite)
+	deployment := complement.Deploy(t, 1)
+	defer deployment.Destroy(t)
+	alice := deployment.Register(t, "hs1", helpers.RegistrationOpts{})
+
+	srv := federation.NewServer(t, deployment,
+		federation.HandleKeyRequests(),
+		federation.HandleTransactionRequests(nil, nil),
+	)
+	cancel := srv.Listen()
+	defer cancel()
+
+	charlie := srv.UserID("charlie")
+
+	roomID := alice.MustCreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+	room := srv.MustJoinRoom(t, deployment, "hs1", roomID, charlie)
+
+	filterID := createFilter(t, alice, map[string]interface{}{
+		"room": map[string]interface{}{
+			"timeline": map[string]interface{}{
+				"limit": 20,
+			},
+		},
+	})
+	_, nextBatch := alice.MustSync(t, client.SyncReq{Filter: filterID})
+	t.Logf("Next batch %s", nextBatch)
+
+	alice.SendEventSynced(t, roomID, b.Event{
+		Type:   "m.room.message",
+		Sender: alice.UserID,
+		Content: map[string]interface{}{
+			"body":    "Hi from Alice!",
+			"msgtype": "m.text",
+		},
+	})
+
+	// Create 50 messages, but don't send them to Alice
+	var missingEvents []gomatrixserverlib.PDU
+	for i := 0; i < 50; i++ {
+		event := srv.MustCreateEvent(t, room, federation.Event{
+			Type:   "m.room.message",
+			Sender: charlie,
+			Content: map[string]interface{}{
+				"body":    "Remote message",
+				"msgtype": "m.text",
+			},
+		})
+		room.AddEvent(event)
+		missingEvents = append(missingEvents, event)
+	}
+
+	// Create one more event that we will send to Alice, which references the
+	// previous 50.
+	lastEvent := srv.MustCreateEvent(t, room, federation.Event{
+		Type:   "m.room.message",
+		Sender: charlie,
+		Content: map[string]interface{}{
+			"body":    "End",
+			"msgtype": "m.text",
+		},
+	})
+	room.AddEvent(lastEvent)
+
+	// Alice's HS will try and fill in the gap, so we need to respond to those
+	// requests.
+	respondToGetMissingEventsEndpoints(t, srv, room, missingEvents)
+
+	srv.MustSendTransaction(t, deployment, "hs1", []json.RawMessage{lastEvent.JSON()}, nil)
+
+	// We now test two different modes of /sync work. The first is when we are
+	// syncing when the server receives the `lastEvent` (and so, at least
+	// Synapse, will start sending down some events immediately).
+	//
+	// The second mode is when we sync *after* all the events have finished
+	// being persisted, and so we get everything in one chunk.
```

Comment:
I'm only kind of understanding the difference here; incremental will have *both* alice and charlie's events in the response, but will be limited? While after full it will be just charlie's?

Reply:
Yes indeed. It's perfectly valid when actively syncing for us to see Alice's message, *then* a limited batch with one or more of charlie's messages. However, we also want to check the case of you syncing only after all the messages come in